### PR TITLE
[Functions] Digest functions #2162

### LINKED ISSUE #2162: [Functions] Digest functions
**Summary**

- [ ] MD5
- [ ] SHA1
- [ ] SHA256

### PR DESCRIPTION
**Pull request by mshauneu**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Add md5, sha1 and sha2

## Changelog

- New Feature

## Related Issues

Fixes #2162

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: website/databend/docs/sqlstatement/hash-functions/md5has.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+---
```
Comment (BohuTANG) [inline website/databend/docs/sqlstatement/hash-functions/md5has.md:1]: Great work!
The document file name is md5hash not md5hash?

Comment (mshauneu) [reply website/databend/docs/sqlstatement/hash-functions/md5has.md]: Sorry, typos. Fixed.

File: common/functions/src/scalars/hashes/sha2hash.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        let l_series = columns[1]
+            .column()
+            .to_minimal_array()?
+            .cast_with_type(&DataType::UInt16)?;
+
+        let opt_iter = i_series
+            .string()?
+            .into_no_null_iter()
+            .zip(l_series.u16()?.into_no_null_iter())
```
Comment (sundy-li) [inline common/functions/src/scalars/hashes/sha2hash.rs:90]: I'm afraid `i_series` and `l_series` may not have the same size.

Eg:
`select sha2( toString(number),  256   )  from numbers(10)`


`l_series` is more likely to be a scalar value DataValue::UInt16, we can have a specific path for this case, thus we can avoid a dynamic `match l` inside the loop to improve the performance.

**Comment by PsiACE** [comment]
Nice work, bro.

File: common/functions/src/scalars/hashes/sha2hash.rs (hunk 2 above)
Comment (mshauneu) [reply common/functions/src/scalars/hashes/sha2hash.rs]: You are right. Changed code to handle scalar case, e.g: `select sha2( toString(number), 256 ) from numbers(10)` for length of hash2.

**Review by sundy-li — APPROVED** [review]

File: common/functions/src/scalars/hashes/sha2hash.rs (hunk 2 above)
Comment (sundy-li) [reply common/functions/src/scalars/hashes/sha2hash.rs]: Great, now it's better yet it's not perfect.

```
 let hash = |i: &[u8], l: &u16| match  ...;
 loop {
      hash(x, y)
 }
```

This did not fit in cpu cache.

Better to use 

```
match l {
    a =>  loop {},
    b => loop {},
     ...
}

```

for the constant case.

But it needs much more duplicated codes.

Comment (sundy-li) [reply common/functions/src/scalars/hashes/sha2hash.rs]: Wow, seems `into_no_null_iter()` did not consider null value case.

**Review by PsiACE — APPROVED** [review]
LGTM

**Pull request merged**